### PR TITLE
chore: bump oxc

### PR DESCRIPTION
**Pull request by Boshen**
### Description

Testing the latest oxc changes cc @overlookmotel

### CODE REVIEW

**Comment by Boshen** [comment]
image

Node is unreliable 🤡

**Comment by overlookmotel** [comment]
@Boshen I don't know how to read these benchmarks. Are they a bit better (they seem to be on CodSpeed) or a bit worse (the "Benchmarks Rust" readings)? Or is "Benchmarks Rust" the NodeJS benchmarks?

**Comment by hyf0** [comment]
> @Boshen I don't know how to read these benchmarks. Are they a bit better (they seem to be on CodSpeed) or a bit worse (the "Benchmarks Rust" readings)? Or is "Benchmarks Rust" the NodeJS benchmarks?

Don't trust benchmarks in ci. Run it locally.

**Comment by Boshen** [comment]
@overlookmotel can you confirm that the usage patterns are all correct before I make a release to crates.io

**Comment by Boshen** [comment]
codspeed benchmark is correct here, there are perf improvements already observed in the oxc repo.

File: crates/rolldown/src/ast_scanner/mod.rs [hunk 1]
```diff
@@ -303,7 +303,7 @@ impl<'me> AstScanner<'me> {
   }
   fn scan_export_default_decl(&mut self, decl: &ExportDefaultDeclaration) {
     let local_binding_for_default_export = match &decl.declaration {
-      oxc::ast::ast::ExportDefaultDeclarationKind::Expression(_exp) => {
+      decl if decl.is_expression() => {
```
Comment (overlookmotel) [inline crates/rolldown/src/ast_scanner/mod.rs]: Change to `oxc::ast::match_expression!(ExportDefaultDeclarationKind) => {`.

(would require `use oxc::ast::ast::ExportDefaultDeclarationKind` higher up)

Also, how do we know that `ExportDefaultDeclarationKind::TSInterfaceDeclaration` and `ExportDefaultDeclarationKind::TSEnumDeclaration` are unreachable here (last arm of match)?

Comment (overlookmotel) [reply crates/rolldown/src/ast_scanner/mod.rs]: Side note: Is there a reason why we don't just import `use oxc::ast::ast::*` at top of the file and remove the repetition of `oxc::ast::ast::...` everywhere?

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 2]
```diff
@@ -48,17 +50,19 @@ impl<'a> SideEffectDetector<'a> {
...
         } || def.value.as_ref().is_some_and(|init| self.detect_side_effect_of_expr(init)))
       }
       ClassElement::AccessorProperty(def) => {
         (match &def.key {
           // FIXME: this is wrong, we should also always check the `def.value`.
-          oxc::ast::ast::PropertyKey::Identifier(_)
-          | oxc::ast::ast::PropertyKey::PrivateIdentifier(_) => false,
-          oxc::ast::ast::PropertyKey::Expression(expr) => self.detect_side_effect_of_expr(expr),
+          PropertyKey::StaticIdentifier(_) | PropertyKey::PrivateIdentifier(_) => false,
+          oxc::ast::match_expression!(PropertyKey) => {
```
Comment (overlookmotel) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Could replace these 2 lines with:

```rs
key @ oxc::ast::match_expression!(PropertyKey) => {
  self.detect_side_effect_of_expr(key.to_expression())
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 3]
```diff
@@ -73,17 +77,18 @@ impl<'a> SideEffectDetector<'a> {
...
     match &member_expr.object {
-      oxc::ast::ast::Expression::Identifier(ident) => {
+      Expression::Identifier(ident) => {
         let object_name = &ident.name;
         // Check if `object_name.prop_name` is pure
         !SIDE_EFFECT_FREE_MEMBER_EXPR_2.contains(&(object_name.as_str(), prop_name.as_str()))
       }
-      oxc::ast::ast::Expression::MemberExpression(mem_expr) => {
-        let MemberExpression::StaticMemberExpression(mem_expr) = &**mem_expr else {
+      oxc::ast::match_member_expression!(Expression) => {
```
Comment (overlookmotel) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Could replace these 2 lines with:

```rs
expr @ oxc::ast::match_member_expression!(Expression) => {
  let mem_expr = expr.to_member_expression();
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 4]
```diff
@@ -114,9 +119,11 @@ impl<'a> SideEffectDetector<'a> {
         obj_expr.properties.iter().any(|obj_prop| match obj_prop {
           oxc::ast::ast::ObjectPropertyKind::ObjectProperty(prop) => {
             let key_side_effect = match &prop.key {
-              oxc::ast::ast::PropertyKey::Identifier(_)
+              oxc::ast::ast::PropertyKey::StaticIdentifier(_)
               | oxc::ast::ast::PropertyKey::PrivateIdentifier(_) => false,
-              oxc::ast::ast::PropertyKey::Expression(expr) => self.detect_side_effect_of_expr(expr),
+              oxc::ast::match_expression!(PropertyKey) => {
```
Comment (overlookmotel) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Could replace these 2 lines with:

```rs
key @ oxc::ast::match_expression!(PropertyKey) => {
  self.detect_side_effect_of_expr(key.to_expression())
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 5]
```diff
@@ -223,19 +232,22 @@ impl<'a> SideEffectDetector<'a> {
...
           // We consider `import ...` has no side effect. However, `import ...` might be rewritten to other statements by the bundler.
           // In that case, we will mark the statement as having side effect in link stage.
           false
         }
         oxc::ast::ast::ModuleDeclaration::ExportDefaultDeclaration(default_decl) => {
+          use oxc::ast::ast::ExportDefaultDeclarationKind;
           match &default_decl.declaration {
-            oxc::ast::ast::ExportDefaultDeclarationKind::Expression(expr) => {
-              self.detect_side_effect_of_expr(expr)
+            oxc::ast::match_expression!(ExportDefaultDeclarationKind) => {
```
Comment (overlookmotel) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Could replace these 2 lines with:

```rs
decl @ oxc::ast::match_expression!(ExportDefaultDeclarationKind) => {
  self.detect_side_effect_of_expr(decl.to_expression())
```

File: crates/rolldown/src/finalizer/impl_visit_mut_for_finalizer.rs [hunk 6]
```diff
@@ -112,7 +112,8 @@ impl<'me, 'ast> VisitMut<'ast> for Finalizer<'me, 'ast> {
           }
         } else if let Some(default_decl) = top_stmt.as_export_default_declaration_mut() {
           match &mut default_decl.declaration {
-            ast::ExportDefaultDeclarationKind::Expression(expr) => {
+            decl if decl.is_expression() => {
```
Comment (overlookmotel) [inline crates/rolldown/src/finalizer/impl_visit_mut_for_finalizer.rs]: Replace with `decl @ match_expression!(ExportDefaultDeclarationKind) => {`.

File: crates/rolldown/src/finalizer/impl_visit_mut_for_finalizer.rs [hunk 7]
```diff
@@ -222,19 +219,18 @@ impl<'me, 'ast> VisitMut<'ast> for Finalizer<'me, 'ast> {
...
-              ast::Declaration::UsingDeclaration(_) => unimplemented!(),
-              _ => {}
-            },
-            ast::Statement::ModuleDeclaration(_) => unreachable!(
+            }
+            ast::Statement::FunctionDeclaration(_) => {
+              fn_stmts.push(stmt);
+            }
+            ast::Statement::UsingDeclaration(_) => unimplemented!(),
+            stmt if stmt.is_module_declaration() => unreachable!(
```
Comment (overlookmotel) [inline crates/rolldown/src/finalizer/impl_visit_mut_for_finalizer.rs]: Replace with `match_module_declaration!(Statement) => unreachable!(`.

File: crates/rolldown_oxc_utils/src/ast_snippet.rs [hunk 8]
```diff
@@ -215,7 +210,7 @@ impl<'ast> AstSnippet<'ast> {
 
     //  __commonJS(...)
     let mut commonjs_call_expr = self.call_expr(commonjs_name);
-    commonjs_call_expr.arguments.push(ast::Argument::Expression(
+    commonjs_call_expr.arguments.push(ast::Argument::from(
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ast_snippet.rs]: You can just create the `Argument::ArrowFunctionExpression` variant directly:

```rs
commonjs_call_expr.arguments.push(
  ast::Argument::ArrowFunctionExpression(arrow_expr.into_in(self.alloc))
);
```

File: crates/rolldown_oxc_utils/src/ast_snippet.rs [hunk 9]
```diff
@@ -246,7 +241,7 @@ impl<'ast> AstSnippet<'ast> {
 
     //  __esm(...)
     let mut commonjs_call_expr = self.call_expr(esm_fn_name);
-    commonjs_call_expr.arguments.push(ast::Argument::Expression(
+    commonjs_call_expr.arguments.push(ast::Argument::from(
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ast_snippet.rs]: You can just create the `Argument::ArrowFunctionExpression` variant directly:

```rs
commonjs_call_expr.arguments.push(
  ast::Argument::ArrowFunctionExpression(arrow_expr.into_in(self.alloc))
);
```

File: crates/rolldown_oxc_utils/src/ast_snippet.rs [hunk 10]
```diff
@@ -302,11 +297,9 @@ impl<'ast> AstSnippet<'ast> {
     id: PassedStr,
     span: Span,
   ) -> ast::AssignmentTarget<'ast> {
-    ast::AssignmentTarget::SimpleAssignmentTarget(
-      ast::SimpleAssignmentTarget::AssignmentTargetIdentifier(
-        self.id_ref(id, span).into_in(self.alloc),
-      ),
-    )
+    ast::AssignmentTarget::from(ast::SimpleAssignmentTarget::AssignmentTargetIdentifier(
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ast_snippet.rs]: You can just create the `AssignmentTarget::AssignmentTargetIdentifier` variant directly:

```rs
ast::AssignmentTarget::AssignmentTargetIdentifier(self.id_ref(id, span).into_in(self.alloc))
```

File: crates/rolldown_oxc_utils/src/ext/mod.rs [hunk 11]
```diff
@@ -80,14 +80,12 @@ impl<'ast> BindingPatternExt<'ast> for ast::BindingPattern<'ast> {
...
-            _ => ast::AssignmentTargetMaybeDefault::AssignmentTarget(
-              binding_pat.into_assignment_target(alloc),
-            ),
+            _ => ast::AssignmentTargetMaybeDefault::from(binding_pat.into_assignment_target(alloc)),
           }));
         });
-        ast::AssignmentTarget::AssignmentTargetPattern(
-          ast::AssignmentTargetPattern::ArrayAssignmentTarget(arr_target.into_in(alloc)),
-        )
+        ast::AssignmentTarget::from(ast::AssignmentTargetPattern::ArrayAssignmentTarget(
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ext/mod.rs]: You can create `AssignmentTarget::ArrayAssignmentTarget` directly:

```rs
ast::AssignmentTarget::ArrayAssignmentTarget(arr_target.into_in(alloc))
```

File: crates/rolldown_oxc_utils/src/ext/mod.rs [hunk 12]
```diff
@@ -116,68 +114,48 @@ pub trait StatementExt<'me, 'ast> {
...
-    )
+    matches!(self, ast::Statement::ImportDeclaration(_))
   }
 
   fn as_import_declaration(&self) -> Option<&ast::ImportDeclaration<'ast>> {
-    if let ast::Statement::ModuleDeclaration(module_decl) = self {
-      if let ast::ModuleDeclaration::ImportDeclaration(import_decl) = &**module_decl {
-        return Some(import_decl);
-      }
+    if let ast::Statement::ImportDeclaration(module_decl) = self {
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ext/mod.rs]: Rename `module_decl` to `import_decl` here for clarity.

File: crates/rolldown_oxc_utils/src/ext/mod.rs [hunk 13]
```diff
@@ -189,11 +167,7 @@ impl<'me, 'ast> StatementExt<'me, 'ast> for ast::Statement<'ast> {
   }
 
   fn as_module_declaration(&self) -> Option<&ast::ModuleDeclaration<'ast>> {
-    if let ast::Statement::ModuleDeclaration(module_decl) = self {
-      Some(module_decl)
-    } else {
-      None
-    }
+    self.as_module_declaration()
```
Comment (overlookmotel) [inline crates/rolldown_oxc_utils/src/ext/mod.rs]: I think we can remove `StatementExt::as_module_declaration` entirely, as `Statement` now provides this method already.

**Comment by overlookmotel** [comment]
> @overlookmotel can you confirm that the usage patterns are all correct before I make a release to crates.io

As far as I can see, everything is correct. I've suggested a bunch of changes, which are for purpose of tidying rather than correctness. They mostly fall into these categories:

1. Remove unnecessary `from` conversions.
2. `foo @ match_*!` pattern.
3. Prefer `foo @ match_bar!()` over `foo if foo.is_bar()`.

The `match!` pattern is not so nice but it does have the benefit of allowing exhaustive matching, which will make it harder for bugs to creep in if we add further enum variants to these types later on. So I think it should be the recommended pattern, even though it's gross.

Please see https://github.com/oxc-project/backlog/issues/15 for a potential alternative which is I think clearer to read, but more "magic".

**Comment by overlookmotel** [comment]
> codspeed benchmark is correct here, there are perf improvements already observed in the oxc repo.

It *should* be a perf improvement as it reduces allocations in some cases.

**Comment by Boshen** [comment]
Got stuck with a peer dependency https://github.com/rolldown/string_wizard/blob/973399661cc7d048002bfb36072fe87eac5565a1/Cargo.toml#L15

**Review by Boshen — COMMENTED** [review]
:shipit:

**Review by hyf0 — APPROVED** [review]
Nice!

**Pull request merged**